Issue 182: Leaderboard changes

woosle1234 opened the pull request:
This pull request for [Issue 182](https://github.com/simon987/Much-Assembly-Required/issues/182):

- Currently, the layout of the leaderboard has changed to show new values, and set up to retrieve the new user stats

- Stats placed in private variables of the Gameobject which is then set and retrieved from the event and listeners for each new userstat

- Added set and get double for userstats to later incorporate time

- Added the events to Cubot plugin for death even and walking event

Working on:

- Still trying to figure out how to get time and set it up

- Also wondering if current configuration and logic works in order to get walk distance and death count

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/DeathCountListener.java:
```diff
@@ -0,0 +1,27 @@
+                    Integer.toString(object.getCounter()));
+
+            ((ControllableUnit) object).getParent().getStats().setInt("deathCount",
+                    DeathCountEvent.getCounter());
```
Here you don't need to pass around the death counter, you can remove the attribute in the Cubot & DeathCountListener, and increment the stats here directly

woosle1234 replied:
Changes made

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/TotalExecutionTimeEvent.java:
```diff
@@ -0,0 +1,29 @@
+
+    }
+
+    public TotalExecutionTimeEvent(GameObject object, double c) {
```
You can confirm in CPU.java but I think that the execution time is an integer. (You can keep the double utility method in UserStats())

woosle1234 replied:
I removed that part completely as I don't see a use for it

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/WalkDistanceListener.java:
```diff
@@ -0,0 +1,27 @@
+            LogManager.LOGGER.info(((ControllableUnit) object).getParent().getUsername() + " Death Count " +
+                    Integer.toString(object.getCounter()));
+
+            ((ControllableUnit) object).getParent().getStats().setInt("walkDistance",
```
Same thing here, you can do +1 here and avoid passing around the walking distance counter variable

woosle1234 replied:
Changed made

simon987 commented on Server/src/main/java/net/simon987/server/web/LeaderBoardPage.java:
```diff
@@ -13,9 +13,12 @@
+        Map<String, Object> model = new HashMap<>(5);
         model.put("session", request.session());
         model.put("stats", GameServer.INSTANCE.getUserStatsHelper().getTopNSetLength("completedVaults", 25));
+        model.put("stats", GameServer.INSTANCE.getUserStatsHelper().getTopNSetLength("deathCount", 25));
```
Here you will have to modify the method so that it provides an array of stats with an attribute for walkDistance, deathCount and completedVaults.
Note that completedVaults itself is already an array of strings - what we're displaying is the number (`.size()` of vaults)

If you need help with that let me know

woosle1234 replied:
I'm not entirely sure how to accomplish this mainly due to the variable type of model and trying to use put with GameServer.INSTANCE.getUserStatsHelper().getTopN("deathCount", 25) giving conflicting variable types

woosle1234 replied:
changed

simon987 commented on Server/src/main/resources/templates/leaderboard.vm:
```diff
@@ -16,13 +16,19 @@
                     <tr>
                         <td>$row.getKey().getUsername()</td>
                         <td>$row.getValue().size()</td>
+                        <td>$row.getValue().size()</td>
```
Same comment as above - I'm calling `size()` to display the number of elements in the array of completed vaults

woosle1234 replied:
would $row.getValue be better or just $row? or should I just use a different name for each stat?

kevinramharak replied:
If i read this correctly value `$stats` is an array containing the top 25 entries where each entry is a key-value pair of type `{ key: User, value: CompletedVault[] }`.

I would suggest keeping the user as the key and modifying the value to be an object containing stats. So the model I am proposing (as pseudo code):
```
interface UserStats {
  CompletedVaults: CompletedVault[];
  ExecutionTime: int;
  Deaths: int;
}
// where
$stats = { key: User, value: UserStats }
```

Then the view template would look like:
```
#foreach($row in $stats)
  <tr>
    <td>$row.getKey().getUsername()</td>
    <td>$row.getValue().CompletedVaults.size()</td>
    <td>$row.getValue().ExecutionTime</td>
    <td>$row.getValue().Deaths</td>
  </tr>
```

This would require you to modify the `getUserStatsHelper().getTopN()` function on the `GameServer.INSTANCE` to retrieve those values and return the correct model.

woosle1234 replied:
I addded a getTopNAll to the UserStatsHelper to get all the stats as an arraylist<map.entry<User,Integer>> but now I'm getting a cast error I don't know why though
![image](https://user-images.githubusercontent.com/10409032/55062367-406afa00-504c-11e9-8c21-a0eb7784872d.png)

kevinramharak replied:
I will spin up your fork and take a look.

kevinramharak replied:
@woosle1234 When i clone your repo the package fails to build with `mvn package`. Any idea why? The tests seem to complete succesfull, but still it reports a test failure in the `Server` package. I am checked out on the `issue-182` branch

```
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] server                                                             [jar]
[INFO] server_root                                                        [pom]
[INFO] plugin-biomassBlob                                                 [jar]
[INFO] plugin-npc                                                         [jar]
[INFO] plugin-misc-hw                                                     [jar]
[INFO] plugin-cubot                                                       [jar]
[INFO] plugin-radioactiveCloud                                            [jar]
[INFO] plugin-construction                                                [jar]
[INFO] 
[INFO] ---------------------< net.simon987.server:server >---------------------
[INFO] Building server 1.4a                                               [1/8]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.2:compile (default-compile) @ server ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ server ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\src\test\resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.2:testCompile (default-testCompile) @ server ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 41 source files to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.7.2:test (default-test) @ server ---
[INFO] Surefire report directory: C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports
The system cannot find the path specified.

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running net.simon987.server.assembly.DWTest
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 2 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 3B 
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.134 sec
Running net.simon987.server.assembly.instruction.AddInstructionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
Running net.simon987.server.assembly.instruction.AndInstructionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.BrkInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetaeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SetaInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetbeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetbInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetccInstructionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec
Running net.simon987.server.assembly.instruction.SetcInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SeteInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetgeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetgInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetleInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetlInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnaeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SetnaInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SetnbeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnbInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
Running net.simon987.server.assembly.instruction.SetncInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetneInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetngeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetngInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnleInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnlInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnoInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnsInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetoInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetsInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetzInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.LabelTest
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 0 bytes (1 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
SEVERE: Unknown mnemonic "999:"@0
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 

Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 2 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 01 
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
Running net.simon987.server.assembly.MemoryTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.OperandTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 sec
Running net.simon987.server.assembly.RegisterSetTest
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.RegisterSet set
INFO: DEBUG: trying to set unknown reg index : 3
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.TestComment
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 0 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 

Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 4 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 3B 00 0C 
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 8 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 53 00 50 00 3D 00 00 
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 sec
Running net.simon987.server.game.objects.ItemsContainerTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.009 sec

Results :

Tests run: 51, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for server_root 1.4a:
[INFO] 
[INFO] server ............................................. FAILURE [  4.792 s]
[INFO] server_root ........................................ SKIPPED
[INFO] plugin-biomassBlob ................................. SKIPPED
[INFO] plugin-npc ......................................... SKIPPED
[INFO] plugin-misc-hw ..................................... SKIPPED
[INFO] plugin-cubot ....................................... SKIPPED
[INFO] plugin-radioactiveCloud ............................ SKIPPED
[INFO] plugin-construction ................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.953 s
[INFO] Finished at: 2019-03-27T18:23:46+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.7.2:test (default-test) on project server: There are test failures.
[ERROR] 
[ERROR] Please refer to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports for the individual test results.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.7.2:test (default-test) on project server: There are test failures.

Please refer to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports for the individual test results.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: There are test failures.

Please refer to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports for the individual test results.
    at org.apache.maven.plugin.surefire.SurefireHelper.reportExecution (SurefireHelper.java:74)
    at org.apache.maven.plugin.surefire.SurefirePlugin.execute (SurefirePlugin.java:642)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

woosle1234 replied:
Not sure I'm running on Windows 10 and works fine

kevinramharak commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/DeathCountEvent.java:
```diff
@@ -0,0 +1,29 @@
+
+    public DeathCountEvent(GameObject object, int c) {
+        setSource(object);
+        if(c>=0){
```
Check https://github.com/simon987/Much-Assembly-Required/blob/master/CONTRIBUTING.md for styling rules.
I don't think it is strictly enforced, but some spacing here would make the core more readable.

Also it would note be much effort to change `c` into `count`, I would prefer a descriptive name over the single letter

woosle1234 replied:
I removed the int c entirely since I found it wasn't being used

woosle1234 commented:
Walking and death counter now works:

1. Upon refresh or server restart the walk distance resets to 0

2. Upon death the walk distance counter resets to 0

3. Walk distance is only set if user achieves a higher walk distance than previously achieved 

Learderboard currently has set values for testing but the logic of code works. Currently facing error when trying to retrieve int value and store within arraylist. Line 41 replicates error:  java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.lang.Integer

Time still need to be set however listener and event setup with proper logic with leaderboard

woosle1234 commented:
> Can you remove this file from the git history?

Removed

kevinramharak commented:
> Walking and death counter now works:
> 
> 1. Upon refresh or server restart the walk distance resets to 0
> 2. Upon death the walk distance counter resets to 0
> 3. Walk distance is only set if user achieves a higher walk distance than previously achieved
> 
> Learderboard currently has set values for testing but the logic of code works. Currently facing error when trying to retrieve int value and store within arraylist. Line 41 replicates error: java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.lang.Integer
> 
> Time still need to be set however listener and event setup with proper logic with leaderboard

I am fairly sure that your cast error has something to do with https://github.com/simon987/Much-Assembly-Required/blob/master/Server/src/main/java/net/simon987/server/user/UserStatsHelper.java#L60.

It seems that in the `getTopNSetLength` (or whatever its name is) function there is a fallback value for if a  stat does not exist. Not sure what its purpose is, but the function seems to operate on the same `MongoCollection<Document>` as the `getTopN` function.

Maybe this is a result of your previous code where you were using the old function to retrieve those stat properties and this line set its default value to an empty `ArrayList`. Then these values where written back to the database. You could confirm by clearing the database of the stored documents.

woosle1234 commented:
> @woosle1234
> 
> > Walking and death counter now works:
> > 
> > 1. Upon refresh or server restart the walk distance resets to 0
> > 2. Upon death the walk distance counter resets to 0
> > 3. Walk distance is only set if user achieves a higher walk distance than previously achieved
> > 
> > Learderboard currently has set values for testing but the logic of code works. Currently facing error when trying to retrieve int value and store within arraylist. Line 41 replicates error: java.lang.ClassCastException: java.util.ArrayList cannot be cast to java.lang.Integer
> > Time still need to be set however listener and event setup with proper logic with leaderboard
> 
> I am fairly sure that your cast error has something to do with https://github.com/simon987/Much-Assembly-Required/blob/master/Server/src/main/java/net/simon987/server/user/UserStatsHelper.java#L60.
> 
> It seems that in the `getTopNSetLength` (or whatever its name is) function there is a fallback value for if a stat does not exist. Not sure what its purpose is, but the function seems to operate on the same `MongoCollection<Document>` as the `getTopN` function.
> 
> Maybe this is a result of your previous code where you were using the old function to retrieve those stat properties and this line set its default value to an empty `ArrayList`. Then these values where written back to the database. You could confirm by clearing the database of the stored documents.

I'll have to test what your saying but it makes sense, either way it should be setup now so that if the stat doesn't exist it should set the integer to 0.

simon987 commented:
Note that the stats are only saved to disk on save ticks (every X ticks the server saves)

woosle1234 commented:
Everything should be working now and implemented:

1. Added a new java file for a Real Life Timer that is constructed at cubot construction

2. I'm basing execution time based on any time where status is not IDLE then adding it to the total execution time of cubot

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/Cubot.java:
```diff
@@ -101,6 +103,16 @@
+    /**
+    *  In game timer
+    */
+    private int time;
```
You don't need to add more attributes to Cubot (See other comments)

woosle1234 replied:
removed

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/Cubot.java:
```diff
@@ -128,6 +140,10 @@ public Cubot(Document document) {
 
+        timer = new RealLifeTimer();
+        timer.run();
+        time = timer.getTime();
```
Not necessary (See other comments)

woosle1234 replied:
removed

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/Cubot.java:
```diff
@@ -157,17 +173,26 @@ public char getMapInfo() {
+            GameEvent event1 = new ExecutionTimeEvent(this, executionTime - time);
+            GameServer.INSTANCE.getEventDispatcher().dispatch(event1);
+        }
+        time = executionTime;
```
Execution time is already calculated, (see comment below)

woosle1234 replied:
removed

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/DeathCountListener.java:
```diff
@@ -0,0 +1,32 @@
+import net.simon987.server.game.objects.GameObject;
+import net.simon987.server.logging.LogManager;
+
+public class DeathCountListener implements GameEventListener {
```
Should be `DeathListener`. You're listening for deaths, not death counts

woosle1234 replied:
done

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/DeathCountListener.java:
```diff
@@ -0,0 +1,32 @@
+        if (object instanceof ControllableUnit) {
+            count = ((ControllableUnit) object).getParent().getStats().getInt("deathCount");
+            count++;
+            LogManager.LOGGER.info(((ControllableUnit) object).getParent().getUsername() + " Death Count " +
```
You can remove this line when you are done debugging

woosle1234 replied:
done

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/ExecutionTimeEvent.java:
```diff
@@ -0,0 +1,23 @@
+import net.simon987.server.event.GameEvent;
+import net.simon987.server.game.objects.GameObject;
+
+public class ExecutionTimeEvent extends GameEvent {
```
You can remove the execution event and its listener. Instead create a function that is called in this loop. There, `cost` is the real time execution, which you can add to the user's total execution time (Without creating an event + a listener)

https://github.com/simon987/Much-Assembly-Required/blob/e378016c566f546d86eebb28892746a7a3fef5e7/Server/src/main/java/net/simon987/server/GameServer.java#L168-L184

Sorry for the confusion, I should have looked into this part a bit more before asking you to make a event listener for this.

woosle1234 replied:
I did what you said I'm not sure if the function works but it is setup and I seem to not have any errors

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/WalkDistanceEvent.java:
```diff
@@ -0,0 +1,17 @@
+import net.simon987.server.event.GameEvent;
+import net.simon987.server.game.objects.GameObject;
+
+public class WalkDistanceEvent extends GameEvent {
```
Should be `WalkEvent`

woosle1234 replied:
done

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/RealLifeTimer.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Timer;
+import java.util.TimerTask;
+
+public class RealLifeTimer{
```
I'm not sure I understand what you're trying to do here
Please add this feature in a separate PR

woosle1234 replied:
I removed this file

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/WalkDistanceListener.java:
```diff
@@ -0,0 +1,39 @@
+import net.simon987.server.logging.LogManager;
+import net.simon987.server.user.UserStats;
+
+public class WalkDistanceListener implements GameEventListener {
```
`WalkListener`

woosle1234 replied:
changed

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/WalkDistanceListener.java:
```diff
@@ -0,0 +1,39 @@
+        WalkDistanceEvent WalkDistanceEvent = (WalkDistanceEvent) event;
+        GameObject object = WalkDistanceEvent.getSource();
+        if (object instanceof ControllableUnit) {
+            //When cubot dies walk counter resets
```
I'd prefer if the walking distance would not reset on death

woosle1234 replied:
changed

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/WalkDistanceListener.java:
```diff
@@ -0,0 +1,39 @@
+                deaths = ((ControllableUnit) object).getParent().getStats().getInt("deathCount");
+            }
+            count++;
+            LogManager.LOGGER.info(((ControllableUnit) object).getParent().getUsername() + " walk distance " +
```
Same comment as above regarding logging

woosle1234 replied:
done

woosle1234 replied:
done

simon987 commented on Plugin Cubot/src/main/java/net/simon987/cubotplugin/event/WalkDistanceListener.java:
```diff
@@ -0,0 +1,39 @@
+            count++;
+            LogManager.LOGGER.info(((ControllableUnit) object).getParent().getUsername() + " walk distance " +
+                    count);
+            //Walk distance is only saved if higher than current hightest distance for user
```
This is not necessary, simply increment the value by one whenever you handle the event. You do not need to store the number in `this.count`

woosle1234 replied:
changed

simon987 commented on Server/src/main/java/net/simon987/server/user/UserStatsHelper.java:
```diff
@@ -38,7 +38,38 @@ public UserStatsHelper(MongoCollection<Document> users) {
+            User user = vaults.get(i).getKey();
+            Map<String, Integer> allStats = new HashMap();
+            allStats.put("completedVaults", vaults.get(i).getValue().size());
+            allStats.put("deathCount", deaths.get(i).getValue());
```
This will not work with multiple users because the order of `deaths` and `vaults` is not necessarily the same.
This part will require a more complex database query and you can leave it as it is for now - I'll create the utility method in UserStats when I get the time

woosle1234 replied:
Okay

woosle1234 commented:
> Hi @woosle1234. I'm very sorry for the delay,
> You did a very good job overall. there's just some tweaks I'd like you to make
> If you need more info just let me know
> 
> Thank you!

No problem!
I took your comments and made changes to the code please let me know if I need to change anything

simon987 commented:
thank you

woosle1234 commented:
> thank you

No problem
